dotnet/Query/Query/NamingStrategy.cs

```diff
@@ -0,0 +1,92 @@
+using System;
+using Io.Cucumber.Messages.Types;
+
+namespace Io.Cucumber.Query
+{
+    /// <summary>
+    /// Names Pickles and other elements in a GherkinDocument.
+    /// </summary>
+    public abstract class NamingStrategy : ILineageReducer<string>
+    {
+        public enum Strategy
+        {
+            LONG,
+            SHORT
+        }
+
+        public enum ExampleName
+        {
+            NUMBER,
+            PICKLE,
+            NUMBER_AND_PICKLE_IF_PARAMETERIZED
+        }
+
+
+        public enum FeatureName
+        {
+            INCLUDE,
+            EXCLUDE
+        }
+
+        public static Builder Create(Strategy strategy)
+        {
+            return new Builder(strategy);
+        }
+
+        public abstract string Reduce(Lineage lineage);
+        public abstract string Reduce(Lineage lineage, Pickle pickle);
+
+        public class Builder
```

Comment:
Here  too: If you can make this work with a constructor with default values, consider doing so.